rubix-bookkeeper/src/test/java/com/qubole/rubix/bookkeeper/TestBookKeeper.java

```diff
@@ -478,19 +485,86 @@ public void testSetAllCacheIngoresCachedBlocks() throws Exception
             TEST_START_BLOCK, TEST_END_BLOCK)
             .setClusterType(ClusterType.TEST_CLUSTER_MANAGER.ordinal())
             .setIncrMetrics(true);
-    bookKeeper.getCacheStatus(request);
+    CacheStatusResponse response = bookKeeper.getCacheStatus(request);
+    int genNumber = response.getGenerationNumber();
 
     long size = bookKeeper.getFileMetadata(TEST_REMOTE_PATH).getCurrentFileSize();
     assertTrue(size == 0, "Non zero size before any data cached: " + size);
-    bookKeeper.setAllCached(TEST_REMOTE_PATH, TEST_FILE_LENGTH, TEST_LAST_MODIFIED, TEST_START_BLOCK, TEST_END_BLOCK);
+    bookKeeper.setAllCached(TEST_REMOTE_PATH, TEST_FILE_LENGTH, TEST_LAST_MODIFIED, TEST_START_BLOCK, TEST_END_BLOCK, genNumber);
     size = bookKeeper.getFileMetadata(TEST_REMOTE_PATH).getCurrentFileSize();
     assertTrue(size == (TEST_END_BLOCK - TEST_START_BLOCK) * CacheConfig.getBlockSize(conf),
             String.format("Expected size: %s but found %s", (TEST_END_BLOCK - TEST_START_BLOCK) * CacheConfig.getBlockSize(conf), size));
 
     // Now a overlapping request of 10 new blocks, new size of FileMetadata should only increase for these 10 new blocks
-    bookKeeper.setAllCached(TEST_REMOTE_PATH, TEST_FILE_LENGTH, TEST_LAST_MODIFIED, TEST_START_BLOCK - 5, TEST_END_BLOCK + 5);
+    bookKeeper.setAllCached(TEST_REMOTE_PATH, TEST_FILE_LENGTH, TEST_LAST_MODIFIED, TEST_START_BLOCK - 5, TEST_END_BLOCK + 5, genNumber);
     long newSize = bookKeeper.getFileMetadata(TEST_REMOTE_PATH).getCurrentFileSize();
     assertTrue(newSize == size + 10 * CacheConfig.getBlockSize(conf),
             String.format("Expected size: %s but found %s", (size + 10) * CacheConfig.getBlockSize(conf), newSize));
   }
+
+  @Test
+  void testReadWithInvalidation() throws IOException, TException, InterruptedException {
```

Comment:
Instead of here, we should add this test in rubix-tests so that we get integration level test. In there we should do it in two parts:
1. Parallel warmup disabled
2. Parallel warmup enabled

In both cases, we need to read via CachingFileSystem while invalidating in parallel.